---
title: "Enrichment Analysis of KOs/KEGG pathways in Predicted Metagenomes"
output: html_notebook
---

# Enrichment Analysis Predicted Metagenomes

A simple enrichment analysis was carried out with the universal enrichment analyzer `enricher` for the KO enrichment analysis.

First, metagenomes were predicted using the `picrust2` pipeline and the original OTU table:

```{bash}
# 2f_otu_table.sample_filtered.no_mitochondria_chloroplast.biom is the OTU table file (from Wallace et al 2018)
# rep_set_conekt.fna is the representative sequences file (from Wallace et al 2018)
picrust2_pipeline.py -i 2f_otu_table.sample_filtered.no_mitochondria_chloroplast.biom -s rep_set_conekt.fna -o picrust2_out_pipeline_not_strat -p 12 --in_traits EC,KO,PFAM
```

`otus_ko.txt` contains the mapping of OTUs to KOs and was generated using the `picrust2_ko_to_enrichment.py` script using the output from unstratified picrust2 (`KO_predicted.tsv`).

```{r}
library(clusterProfiler)

term2genes_df <- read.table("/home/renato/repositories/maize_transcriptome_microbiome_networks/general/otus_ko.txt",
                header = FALSE, sep = "\t")

all_otus_enrichment <- as.vector(read.table("/home/renato/projects/fapesp_bepe_pd/microbiome/co_occurrence_networks/2/day_network_zi_pi_metrics.tsv",
                header = TRUE, sep = "\t"))$Node

day_cross_correlations <- read.table("/home/renato/projects/fapesp_bepe_pd/integration_transcriptome_microbiome/transcriptome_microbiome_paper2025/SparXCC_output_day_common_samples_edgelist.tsv",
                header = TRUE, sep = "\t")
night_cross_correlations <- read.table("/home/renato/projects/fapesp_bepe_pd/integration_transcriptome_microbiome/transcriptome_microbiome_paper2025/SparXCC_output_night_common_samples_edgelist.tsv",
                header = TRUE, sep = "\t")

# KO Enrichment of cross-correlations (OTUs)
day_cross_correlations_ko <- enricher(gene = as.vector(unique(day_cross_correlations$OTU)),
         pvalueCutoff = 0.05,
         pAdjustMethod = "BH",
         universe = all_otus_enrichment,
         TERM2GENE = term2genes_df)

dim(day_cross_correlations_ko[, c("ID", "pvalue", "p.adjust", "qvalue")])

write.table(day_cross_correlations_ko, file = "day_cross_correlations_ko_enrichment.tsv", sep = "\t", quote = FALSE, row.names = FALSE)

night_cross_correlations_ko <- enricher(gene = as.vector(unique(night_cross_correlations$OTU)),
         pvalueCutoff = 0.05,
         pAdjustMethod = "BH",
         universe = all_otus_enrichment,
         TERM2GENE = term2genes_df)

dim(night_cross_correlations_ko[, c("ID", "pvalue", "p.adjust", "qvalue")])

write.table(night_cross_correlations_ko, file = "night_cross_correlations_ko_enrichment.tsv", sep = "\t", quote = FALSE, row.names = FALSE)
```

# Enrichment Analysis Predicted Metagenomes (Pathways excluding Human Diseases)

We noticed that KOs obtained from the predicted metagenomes included a large number of pathways related to human diseases or organimal systems, which is usually associated with multicellular organisms that have organs and tissues.

Therefore, we deleted all pathways related to human diseases, organismal systems, and drug development from the `otus_pathway.txt` file to perform enrichment analysis.

We used:
 * KO - pathway associations from KEGG Database (`ko`)
 * OTU - KEGG pathway association file (`otus_pathway_filtered.txt`) to be used as `term2genes`
 * OTUs from cross-correlations (midday and midnight, separately)

```{r}
library(clusterProfiler)

term2genes_df <- read.table("/home/santosrac/Repositories/maize_transcriptome_microbiome_networks/general/otus_pathway_filtered.txt",
                header = FALSE, sep = "\t")

all_otus_enrichment <- as.vector(read.table("/home/santosrac/Repositories/maize_transcriptome_microbiome_networks/day_network_zi_pi_metrics.tsv",
                header = TRUE, sep = "\t"))$Node

day_cross_correlations <- read.table("/home/santosrac/Projects/UGA_RACS/IntegrationMetataxExpression/SparXCC_OTUs_paper/SparXCC_output_day_common_samples_edgelist.tsv",
                header = TRUE, sep = "\t")
night_cross_correlations <- read.table("/home/santosrac/Projects/UGA_RACS/IntegrationMetataxExpression/SparXCC_OTUs_paper/SparXCC_output_night_common_samples_edgelist.tsv",
                header = TRUE, sep = "\t")

# Kegg Pathway Enrichment of cross-correlations (OTUs)
day_cross_correlations_pathways <- enricher(gene = as.vector(unique(day_cross_correlations$OTU)),
         pvalueCutoff = 0.05,
         pAdjustMethod = "BH",
         universe = all_otus_enrichment,
         TERM2GENE = term2genes_df)

dim(day_cross_correlations_pathways[, c("ID", "pvalue", "p.adjust", "qvalue")])

day_cross_correlations_pathways[, c("ID", "pvalue", "p.adjust", "qvalue")]

night_cross_correlations_pathways <- enricher(gene = as.vector(unique(night_cross_correlations$OTU)),
         pvalueCutoff = 0.05,
         pAdjustMethod = "BH",
         universe = all_otus_enrichment,
         TERM2GENE = term2genes_df)

dim(night_cross_correlations_pathways[, c("ID", "pvalue", "p.adjust", "qvalue")])

night_cross_correlations_pathways[, c("ID", "pvalue", "p.adjust", "qvalue")]
```