---
title: "Co-expression network analysis"
output: html_notebook
---

This notebook describes data exploratory analysis and co-expression network analysis of RNA-seq data.

At least part of the analyses from the [SimpleTidy GeneCoEx workflow](https://github.com/cxli233/SimpleTidy_GeneCoEx) (created by Dr. Chenxin Li, at the University of Georgia) were used in this notebook.


# Importing TPM gene expression matrices and initial data exploration

Importing the necessary libraries for the analysis:

```{r}
library(tidyverse)
library(igraph)
library(ggraph)
library(readxl)
library(patchwork)
library(RColorBrewer)
library(viridis)
```

Importing the TPM gene expression matrices.

`Exp_table` was generated by running Salmon on the RNA-seq data from Kremling et al. (2018) and using the index of Maize v5 as the reference transcriptome.
Expression counts were normalized to TPM and genes were filtered out by expression abundance (RPKM) and coefficient of variation (CV).

A second matrix, `Full_Expr_table`, was generated using the same pipeline but without any filtering steps.

```{r}
Full_Expr_table <- read_tsv("/home/renato/projects/fapesp_bepe_pd/transcriptome/kremling_expression_v5_tpm.tsv", col_types = cols())
Exp_table <- read_tsv("/home/renato/projects/fapesp_bepe_pd/transcriptome/kremling_expression_v5_tpm_filtered_cv_filtered.tsv", col_types = cols())
head(Full_Expr_table)
dim(Full_Expr_table)
head(Exp_table)
dim(Exp_table)
```

Generating the Tidy data frame for the gene expression matrix:

```{r}
Full_Expr_table_long <- Full_Expr_table %>% 
  pivot_longer(cols = !Name, names_to = "SampleName", values_to = "TPM") %>% 
  mutate(logTPM = log10(TPM + 1))

head(Full_Expr_table_long)

Exp_table_long <- Exp_table %>% 
  pivot_longer(cols = !Name, names_to = "SampleName", values_to = "TPM") %>% 
  mutate(logTPM = log10(TPM + 1))

head(Exp_table_long)
```

Generating a wide version again (with log TPM values):

```{r}
Exp_table_log_wide <- Exp_table_long %>% 
  select(Name, SampleName, logTPM) %>% 
  pivot_wider(names_from = SampleName, values_from = logTPM)

head(Exp_table_log_wide)

Full_Expr_table_long_wide <- Full_Expr_table_long %>% 
  select(Name, SampleName, logTPM) %>% 
  pivot_wider(names_from = SampleName, values_from = logTPM)

head(Full_Expr_table_long_wide)
```

Importing metadata:

```{r}
sample_metadata <- read_tsv("/home/renato/projects/fapesp_bepe_pd/transcriptome/sample_annotation.txt",
                            show_col_types = FALSE)
head(sample_metadata)
```


## Principal Component Analysis (PCA)

To make sure the experimental design reflects the comparison we intend to make in our analyses (day and night contrast), a PCA was performed on both matrices (the full and the filtered gene sets).

```{r}
my_pca <- prcomp(t(Exp_table_log_wide[, -1]))
pca_importance <- as.data.frame(t(summary(my_pca)$importance))
head(pca_importance, 20)

my_pca_full <- prcomp(t(Full_Expr_table_long_wide[, -1]))
pca_importance_full <- as.data.frame(t(summary(my_pca_full)$importance))
head(pca_importance_full, 20)
```


Adding metadata to the PCA matrix:

```{r}
PCA_coord_full <- my_pca_full$x[, 1:10] %>% 
  as.data.frame() %>% 
  mutate(SampleName = row.names(.)) %>% 
  full_join(sample_metadata %>% 
              select(Genotype, Subpopulation, DayPeriod, Tissue, SampleName), by = "SampleName")

head(PCA_coord_full)

PCA_coord <- my_pca$x[, 1:10] %>% 
  as.data.frame() %>% 
  mutate(SampleName = row.names(.)) %>% 
  full_join(sample_metadata %>% 
              select(Genotype, Subpopulation, DayPeriod, Tissue, SampleName), by = "SampleName")

head(PCA_coord)
```

Plotting the PCA with the Full RNA-Seq matrix (log transformed TPM values):

```{r}
PCA_by_day_period_full <- PCA_coord_full %>% 
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(fill = DayPeriod), color = "grey20", shape = 21, size = 3, alpha = 0.8) +
  scale_fill_manual(values = brewer.pal(n = 3, "Accent")) +
  labs(x = paste("PC1 (", pca_importance_full[1, 2] %>% signif(3)*100, "% of Variance)", sep = ""), 
       y = paste("PC2 (", pca_importance_full[2, 2] %>% signif(3)*100, "% of Variance)", "  ", sep = ""),
       fill = NULL) +  
  theme_bw() +
  theme(
    text = element_text(size= 14),
    axis.text = element_text(color = "black")
  )

PCA_by_day_period_full
```

Plotting the PCA with the RNA-Seq matrix (log transformed TPM values) for the subset filtered genes used in some of the co-expression analyses and the cross-correlations (with microbiome):

```{r}
PCA_by_day_period <- PCA_coord %>% 
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(fill = DayPeriod), color = "grey20", shape = 21, size = 3, alpha = 0.8) +
  scale_fill_manual(values = brewer.pal(n = 3, "Accent")) +
  labs(x = paste("PC1 (", pca_importance[1, 2] %>% signif(3)*100, "% of Variance)", sep = ""), 
       y = paste("PC2 (", pca_importance[2, 2] %>% signif(3)*100, "% of Variance)", "  ", sep = ""),
       fill = NULL) +  
  theme_bw() +
  theme(
    text = element_text(size= 14),
    axis.text = element_text(color = "black")
  )

PCA_by_day_period
```

The same plot was also generated with the maize subpopulations:

```{r}
PCA_by_subpopulation <- PCA_coord %>% 
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(fill = Subpopulation), color = "grey20", shape = 21, size = 3, alpha = 0.8) +
  labs(x = paste("PC1 (", pca_importance[1, 2] %>% signif(3)*100, "% of Variance)", sep = ""), 
       y = paste("PC2 (", pca_importance[2, 2] %>% signif(3)*100, "% of Variance)", "  ", sep = ""),
       fill = NULL) +  
  theme_bw() +
  theme(
    text = element_text(size= 14),
    axis.text = element_text(color = "black")
  )

PCA_by_subpopulation
```

