---
title: "Differential Abundance Analysis on the day periods"
output: html_notebook
---

# Differential Abundance Analysis (DAA)

Importing required libraries:

```{r}
library(DT)
library(tidyverse)
library(ANCOMBC)
library(phyloseq)
library(edgeR)
library(RColorBrewer)
library(scico)
```

#Importing data and running ancombc with our data:

# Importing day and night count matrices (filtered by rel. abundance and coefficient of variation)
day_otu_filtered_sparxcc_wallace2018 <- read.csv('filtered_otu_table_day_filtered_rel_abund_cv_filtered.tsv', sep='\t', header=TRUE, row.names=1)

#day_otu_filtered_sparxcc_wallace2018 <- read.csv('/home/renato/projects/fapesp_bepe_pd/microbiome/filtered_otu_table_day_filtered_rel_abund_cv_filtered.tsv', sep='\t', header=TRUE, row.names=1)
colnames(day_otu_filtered_sparxcc_wallace2018) <- gsub("^X", "", colnames(day_otu_filtered_sparxcc_wallace2018))
colnames(day_otu_filtered_sparxcc_wallace2018) <- gsub("^", "day_", colnames(day_otu_filtered_sparxcc_wallace2018))
night_otu_filtered_sparxcc_wallace2018 <- read.csv('filtered_otu_table_night_filtered_rel_abund_cv_filtered.tsv', sep='\t', header=TRUE, row.names=1)
#night_otu_filtered_sparxcc_wallace2018 <- read.csv('/home/renato/projects/fapesp_bepe_pd/microbiome/filtered_otu_table_night_filtered_rel_abund_cv_filtered.tsv', sep='\t', header=TRUE, row.names=1)
colnames(night_otu_filtered_sparxcc_wallace2018) <- gsub("^X", "", colnames(night_otu_filtered_sparxcc_wallace2018))
colnames(night_otu_filtered_sparxcc_wallace2018) <- gsub("^", "night_", colnames(night_otu_filtered_sparxcc_wallace2018))
head(day_otu_filtered_sparxcc_wallace2018)
head(night_otu_filtered_sparxcc_wallace2018)

# Merging counts matrices
merged_otu_filtered_sparxcc_wallace2018 <- merge(day_otu_filtered_sparxcc_wallace2018, night_otu_filtered_sparxcc_wallace2018, by = "row.names", all = TRUE)
rownames(merged_otu_filtered_sparxcc_wallace2018) <- merged_otu_filtered_sparxcc_wallace2018$Row.names
merged_otu_filtered_sparxcc_wallace2018 <- merged_otu_filtered_sparxcc_wallace2018[ , -1]
dim(merged_otu_filtered_sparxcc_wallace2018)
head(merged_otu_filtered_sparxcc_wallace2018)

metadata <- read.csv('sample_annotation.txt', sep='\t', header=TRUE, row.names=1)
#metadata <- read.csv('/home/renato/repositories/maize_transcriptome_microbiome_networks/sample_annotation.txt', sep='\t', header=TRUE, row.names=1)
dim(metadata)
head(metadata)
# Removing substring from the beginning of row names
rownames(metadata) <- sub("^exp_", "", rownames(metadata))

# Creating a phyloseq object without a tax table
otu_table_obj <- otu_table(as.matrix(merged_otu_filtered_sparxcc_wallace2018), taxa_are_rows = TRUE)
sample_data_obj <- sample_data(metadata)
phseq <- phyloseq(otu_table_obj, sample_data_obj)

set.seed(123)
output = ancombc(data = phseq,
                  tax_level = NULL,
                  formula = "Subpopulation + DayPeriod",
                  p_adj_method = "holm",
                  prv_cut = 0,
                  lib_cut = 1000,
                  group = "DayPeriod",
                  struc_zero = FALSE,
                  neg_lb = TRUE,
                  alpha = 0.05,
                  global = TRUE,
                  verbose = TRUE)

res = output$res
res_global = output$res_global

# Checking if rows in output$feature_table are exactly equal to otu_table_obj
all.equal(as.data.frame(output$feature_table), as.data.frame(merged_otu_filtered_sparxcc_wallace2018))

tab_lfc = res$lfc
head(tab_lfc)
colnames(tab_lfc)
col_name = c("Taxon", "Intercept", "Subpopulation Mixed", "Subpopulation Non-stiff stalk", "Subpopulation Popcorn", 
             "Subpopulation Stiff stalk", "Subpopulation Sweet", "Subpopulation Tropical-subtropical", "Subpopulation unclassified",
             "DayPeriod Night")
             # Additional columns for results if we consider the interaction (subpopulation and day period)
             #, "Subpopulationmixed:DayPeriodnight", "Subpopulationnon-stiff stalk:DayPeriodnight", "Subpopulationpopcorn:DayPeriodnight",
             #"Subpopulationstiff stalk:DayPeriodnight", "Subpopulationsweet:DayPeriodnight", "Subpopulationtropical-subtropical:DayPeriodnight",
             #"Subpopulationunclassified:DayPeriodnight")
colnames(tab_lfc) = col_name
aux1 <- tab_lfc %>% 
  datatable(caption = "Log Fold Changes from the Primary Result") %>%
  formatRound(col_name[-1], digits = 2)

tab_diff = res$diff_abn
colnames(tab_diff) = col_name
aux2 <- tab_diff %>% 
  datatable(caption = "Differentially Abundant Taxa from the Primary Result")

(res$q_val < 0.05 & res$diff_abn == TRUE)

# Select only the taxa considered differentially abundant comparing two Day Periods
(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')]

# Creating a new dataframe with specific columns from filtered ancombc results
filtered_results <- data.frame(
  Taxon = rownames(output$feature_table[(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')],]),
  LogFoldChange = res$lfc[(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')], c("DayPeriodnight")],
  PValue = res$p_val[(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')], c("DayPeriodnight")],
  AdjustedPValue = res$q_val[(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')], c("DayPeriodnight")],
  DifferentialAbundance = res$diff_abn[(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')], c("DayPeriodnight")]
)

dim(filtered_results)
head(filtered_results)
tail(filtered_results)

sum(filtered_results$LogFoldChange > 0, na.rm = TRUE)
sum(filtered_results$LogFoldChange < 0, na.rm = TRUE)

filtered_results_sorted_lfc <- filtered_results[order(filtered_results$LogFoldChange, decreasing = TRUE), ]
head(filtered_results_sorted_lfc, n = 10)
tail(filtered_results_sorted_lfc, n = 10)

write.csv(filtered_results_sorted_lfc, file = "filtered_results_sorted_lfc.csv", row.names = FALSE)

# Import the GTDB taxonomy table
gtdb_taxonomy <- read.csv("gtdb_taxonomy.tsv", sep='\t', header=TRUE)
gtdb_taxonomy <- gtdb_taxonomy[, -3]
rownames(gtdb_taxonomy) <- gtdb_taxonomy$Feature.ID
head(gtdb_taxonomy)

# Selecting OTUs with positive fold changes
positive_fold_change_otus <- filtered_results %>%
  filter(LogFoldChange > 0) %>%
  select(Taxon, LogFoldChange, AdjustedPValue)

head(positive_fold_change_otus)

# Selecting OTUs with negative fold changes
negative_fold_change_otus <- filtered_results %>%
  filter(LogFoldChange < 0) %>%
  select(Taxon, LogFoldChange, AdjustedPValue)

head(negative_fold_change_otus)

# Merging positive_fold_change_otus with gtdb_taxonomy to get taxonomy information
positive_fold_change_otus_with_taxonomy <- positive_fold_change_otus %>%
  left_join(gtdb_taxonomy, by = c("Taxon" = "Feature.ID"))

head(positive_fold_change_otus_with_taxonomy)
dim(positive_fold_change_otus_with_taxonomy)
positive_fold_change_otus_with_taxonomy[order(positive_fold_change_otus_with_taxonomy$LogFoldChange, decreasing = TRUE), ]

# Merging negative_fold_change_otus with gtdb_taxonomy to get taxonomy information
negative_fold_change_otus_with_taxonomy <- negative_fold_change_otus %>%
  left_join(gtdb_taxonomy, by = c("Taxon" = "Feature.ID"))
head(negative_fold_change_otus_with_taxonomy)
dim(negative_fold_change_otus_with_taxonomy)
negative_fold_change_otus_with_taxonomy[order(negative_fold_change_otus_with_taxonomy$LogFoldChange, decreasing = FALSE), ]

# Merge positive and negative fold change OTUs with taxonomy into a single dataframe
merged_fold_change_otus_with_taxonomy <- bind_rows(
  positive_fold_change_otus_with_taxonomy %>% mutate(FoldChangeDirection = "Positive"),
  negative_fold_change_otus_with_taxonomy %>% mutate(FoldChangeDirection = "Negative")
)

head(merged_fold_change_otus_with_taxonomy)
dim(merged_fold_change_otus_with_taxonomy)

write.csv(merged_fold_change_otus_with_taxonomy, file = "merged_fold_change_otus_with_taxonomy.csv", row.names = FALSE)

# Computing CPM for OTU counts
otu <- as(otu_table(phseq), "matrix")
otu_cpm <- cpm(otu, log = FALSE)

# Creating a new phyloseq object with the CPM values
otu_table(phseq) <- otu_table(otu_cpm, taxa_are_rows = TRUE)
phsig <- prune_taxa(filtered_results$Taxon, phseq)

library(ggplot2)

df_plot <- psmelt(phsig)
head(df_plot)

# Computing average OTU abundance for each group in DayPeriod
average_otu <- df_plot %>%
  group_by(DayPeriod, OTU) %>%
  summarise(AverageAbundance = mean(Abundance, na.rm = TRUE)) %>%
  ungroup()

head(average_otu)

# Compute fold change of a particular OTU between two groups in DayPeriod
otu_of_interest <- "305420"  # Replace with the OTU of interest
fold_change <- average_otu %>%
  filter(OTU == otu_of_interest) %>%
  spread(key = DayPeriod, value = AverageAbundance) %>%
  mutate(FoldChange = night / day)  # Assuming "Day" and "Night" are the group names

fold_change

# Boxplot or violin plot
ggplot(df_plot, aes(x = DayPeriod, y = Abundance, fill = DayPeriod)) +
  geom_boxplot(outlier.shape = NA, alpha = 0.6) +
  geom_jitter(width = 0.2, size = 0.8, alpha = 0.4) +
  facet_wrap(~ OTU, scales = "free_y") +
  labs(title = "Differentially Abundant OTUs between Day and Night",
       y = "Abundance (CPM)", x = "") +
  theme_minimal() +
  theme(strip.text = element_text(size = 10), axis.text.x = element_text(angle = 45, hjust = 1))

sample_info <- data.frame(sample_data(phseq))
#annotation_col <- data.frame(DayPeriod = sample_info$DayPeriod, Subpopulation = sample_info$Subpopulation)
annotation_col <- data.frame(DayPeriod = sample_info$DayPeriod)

rownames(annotation_col) <- rownames(sample_info)

library(pheatmap)

# Function to get column order for each group
get_group_col_order <- function(mat, group_labels) {
    unique_groups <- unique(group_labels)
    ordered_cols <- c()
    for (grp in unique_groups) {
        grp_cols <- which(group_labels == grp)
        if (length(grp_cols) > 1) {
            # Cluster columns within group
            d <- dist(t(mat[, grp_cols, drop = FALSE]))
            hc <- hclust(d)
            ordered_grp_cols <- grp_cols[order.dendrogram(as.dendrogram(hc))]
        } else {
            ordered_grp_cols <- grp_cols
        }
        ordered_cols <- c(ordered_cols, ordered_grp_cols)
    }
    return(ordered_cols)
}

col_order <- get_group_col_order(otu_table(phseq), annotation_col$DayPeriod)
heatmap_mat <- otu_table(phseq)[, col_order]
annotation_col <- annotation_col[col_order, , drop = FALSE]
colorSubpopulations = brewer.pal(n = 8, name = "Set2")
ann_colors = list(
    DayPeriod = c(day = "#FFD700", night = "#1E3A8A")
#    Subpopulation = colorSubpopulations
    )
#names(ann_colors$Subpopulation) <- levels(as.factor(sample_info$Subpopulation))
breaksList = seq(-2, 2, by = 0.1)

p <- pheatmap(heatmap_mat[filtered_results$Taxon,],
      annotation_col = annotation_col,
      legend_title = "Day Period",
      cluster_cols = FALSE,  # Disable clustering for columns to preserve group order
      cluster_rows = TRUE,
      gaps_row = F,
      gaps_cols = F,
      cutree_rows = 10,
      border_color = NA,
      scale = "row",  # z-score normalization across each OTU
      show_rownames = FALSE,
      show_colnames = FALSE,
#      fontsize = 12,  
#      fontsize_row =6,
      legend = TRUE,
      cellwidth = 0.5,
      cellheight = 2*2,
      color = scico(length(breaksList), palette = 'vikO'),
      breaks = breaksList,
      annotation_colors = ann_colors,
      silent = F,
      main = "Differentially Abundant OTUs (log2 CPM)")

grob <- p$gtable

pdf("heatmap_differentially_abundant_otus.pdf")
grid::grid.newpage()
grid::grid.draw(grob)
dev.off()

png("heatmap_differentially_abundant_otus.png", width = 1600, height = 2100, res = 300)
grid::grid.newpage()
grid::grid.draw(grob)
dev.off()

### HEATMAP GROUPS OF OUTS

df <- heatmap_mat[filtered_results$Taxon,]

tmp <- pheatmap(df, cluster_rows = T, cluster_cols = T, scale = "row")
clusters_otus <- as.data.frame(cutree(tmp$tree_row, k = 10))
colnames(clusters_otus) <- "cluster"

get_eigen <- function(module_no){
	otu <- row.names(clusters_otus %>% filter(cluster == module_no))
        mat <- df[otu,]
	pca <- prcomp(t(mat))
	eigen <-pca$x[,1]

	# Calculate module eigengene as the mean expression across all genes in the module
        module_mean <- colMeans(mat, na.rm = TRUE)
        # Check correlation between PC1 and module mean
        correlation <- cor(eigen, module_mean)
        # If correlation is negative, flip the eigengene
	if(correlation < 0) {
                eigen <- -eigen
        
	}
	row <- data.frame(eigen = t(eigen), row.names = paste0("Cluster_",module_no))
	colnames(row) <- names(eigen)
	return(row)
}

results_list <- list()
loop_modules <- levels(as.factor(clusters_otus$cluster))
for (module in loop_modules){
	result <- get_eigen(module)
        results_list[[as.character(module)]] <- result
	cat("Module:", module, "\n")
}

results_df <- bind_rows(results_list, .id = "module")

eigen_mat <- results_df %>%
        select(-module) %>%
        as.matrix()

ann_colors = list(DayPeriod = c(day = "#FFD700", night = "#1E3A8A"))

#sample_info <- data.frame(SampleName = colnames(results_df))
#sample_info$Period <- ifelse(grepl("exp_day", sample_info$SampleName, ignore.case = TRUE), "Day", "Night")
#row.names(sample_info) <- sample_info$SampleName
#sample_info$SampleName <- NULL

breaksList = seq(-2, 2, by = 0.1)

p <- pheatmap(eigen_mat,
	      main = "PUT SOMETHING HERE",
	      scale = "row",
	      annotation_col = annotation_col,
	      show_rownames = T,
	      show_colnames = F,
	      breaks = breaksList,
	      annotation_colors = ann_colors,
	      col = scico(length(breaksList), palette = 'vikO'),
	      cluster_cols = F,
	      cluster_rows = T,
	      cellwidth = 1,
	      cellheight = 9)

grob <- p$gtable

pdf("heatmap_clusters_otus.pdf")
grid::grid.newpage()
grid::grid.draw(grob)
dev.off()

png("heatmap_clusters_otus.png", res = 300, width = 2.4*1200, height = 2500)
grid::grid.newpage()
grid::grid.draw(grob)
dev.off()


