---
title: "Differential Abundance Analysis on the day periods"
output: html_notebook
---

# Differential Abundance Analysis (DAA)

Installing ANCOM and other required packages:

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("ANCOMBC")

# Getting the microbiome package (some of the test data is from there):
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("microbiome")

install.packages("BiocManager")
BiocManager::install("phyloseq")
install.packages("tidyverse")
install.packages("DT")

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("edgeR")
```

Importing required libraries:

```{r}
library(DT)
library(tidyverse)
library(ANCOMBC)
library(phyloseq)
library(edgeR)
```

Testing ancom using example dataset from the Microbiome package:

```{r}
data(atlas1006, package = "microbiome")
pseq = phyloseq::subset_samples(atlas1006, time == 0)
meta_data = microbiome::meta(pseq)
meta_data$bmi = recode(meta_data$bmi_group,
                       obese = "obese",
                       severeobese = "obese",
                       morbidobese = "obese")
meta_data                       
meta_data$bmi = factor(meta_data$bmi, levels = c("obese", "overweight", "lean"))
meta_data$region = recode(as.character(meta_data$nationality),
                          Scandinavia = "NE", UKIE = "NE", SouthEurope = "SE", 
                          CentralEurope = "CE", EasternEurope = "EE",
                          .missing = "unknown")

phyloseq::sample_data(pseq) = meta_data

# Subset to lean, overweight, and obese subjects
pseq = phyloseq::subset_samples(pseq, bmi %in% c("lean", "overweight", "obese"))
# Discard "EE" as it contains only 1 subject
# Discard subjects with missing values of region
pseq = phyloseq::subset_samples(pseq, ! region %in% c("EE", "unknown"))

out = ancombc(data = pseq, tax_level = "Family", 
              formula = "age + region + bmi", 
              p_adj_method = "holm", prv_cut = 0.10, lib_cut = 1000, 
              group = "bmi", struc_zero = TRUE, neg_lb = TRUE, tol = 1e-5, 
              max_iter = 100, conserve = TRUE, alpha = 0.05, global = TRUE,
              n_cl = 1, verbose = TRUE)

res = out$res
res_global = out$res_global
```


Importing data and running ancombc with our data:

```{r}
# Importing day and night count matrices (filtered by rel. abundance and coefficient of variation)
day_otu_filtered_sparxcc_wallace2018 <- read.csv('/home/santosrac/Projects/UGA_RACS/IntegrationMetataxExpression/SparXCC_OTUs_paper/filtered_otu_table_day_filtered_rel_abund_cv_filtered.tsv', sep='\t', header=TRUE, row.names=1)
#day_otu_filtered_sparxcc_wallace2018 <- read.csv('/home/renato/projects/fapesp_bepe_pd/microbiome/filtered_otu_table_day_filtered_rel_abund_cv_filtered.tsv', sep='\t', header=TRUE, row.names=1)
colnames(day_otu_filtered_sparxcc_wallace2018) <- gsub("^X", "", colnames(day_otu_filtered_sparxcc_wallace2018))
colnames(day_otu_filtered_sparxcc_wallace2018) <- gsub("^", "day_", colnames(day_otu_filtered_sparxcc_wallace2018))
night_otu_filtered_sparxcc_wallace2018 <- read.csv('/home/santosrac/Projects/UGA_RACS/IntegrationMetataxExpression/SparXCC_OTUs_paper/filtered_otu_table_night_filtered_rel_abund_cv_filtered.tsv', sep='\t', header=TRUE, row.names=1)
#night_otu_filtered_sparxcc_wallace2018 <- read.csv('/home/renato/projects/fapesp_bepe_pd/microbiome/filtered_otu_table_night_filtered_rel_abund_cv_filtered.tsv', sep='\t', header=TRUE, row.names=1)
colnames(night_otu_filtered_sparxcc_wallace2018) <- gsub("^X", "", colnames(night_otu_filtered_sparxcc_wallace2018))
colnames(night_otu_filtered_sparxcc_wallace2018) <- gsub("^", "night_", colnames(night_otu_filtered_sparxcc_wallace2018))
head(day_otu_filtered_sparxcc_wallace2018)
head(night_otu_filtered_sparxcc_wallace2018)

# Merging counts matrices
merged_otu_filtered_sparxcc_wallace2018 <- merge(day_otu_filtered_sparxcc_wallace2018, night_otu_filtered_sparxcc_wallace2018, by = "row.names", all = TRUE)
rownames(merged_otu_filtered_sparxcc_wallace2018) <- merged_otu_filtered_sparxcc_wallace2018$Row.names
merged_otu_filtered_sparxcc_wallace2018 <- merged_otu_filtered_sparxcc_wallace2018[ , -1]
dim(merged_otu_filtered_sparxcc_wallace2018)
head(merged_otu_filtered_sparxcc_wallace2018)

metadata <- read.csv('/home/santosrac/Repositories/maize_transcriptome_microbiome_networks/sample_annotation.txt', sep='\t', header=TRUE, row.names=1)
#metadata <- read.csv('/home/renato/repositories/maize_transcriptome_microbiome_networks/sample_annotation.txt', sep='\t', header=TRUE, row.names=1)
dim(metadata)
head(metadata)
# Removing substring from the beginning of row names
rownames(metadata) <- sub("^exp_", "", rownames(metadata))

# Creating a phyloseq object without a tax table
otu_table_obj <- otu_table(as.matrix(merged_otu_filtered_sparxcc_wallace2018), taxa_are_rows = TRUE)
sample_data_obj <- sample_data(metadata)
phseq <- phyloseq(otu_table_obj, sample_data_obj)

set.seed(123)
output = ancombc(data = phseq,
                  tax_level = NULL,
                  formula = "Subpopulation + DayPeriod",
                  p_adj_method = "holm",
                  prv_cut = 0,
                  lib_cut = 1000,
                  group = "DayPeriod",
                  struc_zero = FALSE,
                  neg_lb = TRUE,
                  alpha = 0.05,
                  global = TRUE,
                  verbose = TRUE)

res = output$res
res_global = output$res_global

# Checking if rows in output$feature_table are exactly equal to otu_table_obj
all.equal(as.data.frame(output$feature_table), as.data.frame(merged_otu_filtered_sparxcc_wallace2018))

tab_lfc = res$lfc
head(tab_lfc)
colnames(tab_lfc)
col_name = c("Taxon", "Intercept", "Subpopulation Mixed", "Subpopulation Non-stiff stalk", "Subpopulation Popcorn", 
             "Subpopulation Stiff stalk", "Subpopulation Sweet", "Subpopulation Tropical-subtropical", "Subpopulation unclassified",
             "DayPeriod Night")
             # Additional columns for results if we consider the interaction (subpopulation and day period)
             #, "Subpopulationmixed:DayPeriodnight", "Subpopulationnon-stiff stalk:DayPeriodnight", "Subpopulationpopcorn:DayPeriodnight",
             #"Subpopulationstiff stalk:DayPeriodnight", "Subpopulationsweet:DayPeriodnight", "Subpopulationtropical-subtropical:DayPeriodnight",
             #"Subpopulationunclassified:DayPeriodnight")
colnames(tab_lfc) = col_name
tab_lfc %>% 
  datatable(caption = "Log Fold Changes from the Primary Result") %>%
  formatRound(col_name[-1], digits = 2)

tab_diff = res$diff_abn
colnames(tab_diff) = col_name
tab_diff %>% 
  datatable(caption = "Differentially Abundant Taxa from the Primary Result")

(res$q_val < 0.05 & res$diff_abn == TRUE)

# Select only the taxa considered differentially abundant comparing two Day Periods
(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')]

# Creating a new dataframe with specific columns from filtered ancombc results
filtered_results <- data.frame(
  Taxon = rownames(output$feature_table[(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')],]),
  LogFoldChange = res$lfc[(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')], c("DayPeriodnight")],
  PValue = res$p_val[(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')], c("DayPeriodnight")],
  AdjustedPValue = res$q_val[(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')], c("DayPeriodnight")],
  DifferentialAbundance = res$diff_abn[(res$q_val < 0.05 & res$diff_abn == TRUE)[, c('DayPeriodnight')], c("DayPeriodnight")]
)

dim(filtered_results)
head(filtered_results)
tail(filtered_results)

sum(filtered_results$LogFoldChange > 0, na.rm = TRUE)
sum(filtered_results$LogFoldChange < 0, na.rm = TRUE)

filtered_results_sorted_lfc <- filtered_results[order(filtered_results$LogFoldChange, decreasing = TRUE), ]
head(filtered_results_sorted_lfc, n = 10)
tail(filtered_results_sorted_lfc, n = 10)

# Import the GTDB taxonomy table
gtdb_taxonomy <- read.csv("/home/santosrac/Projects/UGA_RACS/16S/otu_taxonomy/gtdb_taxonomy.tsv", sep='\t', header=TRUE)
gtdb_taxonomy <- gtdb_taxonomy[, -3]
rownames(gtdb_taxonomy) <- gtdb_taxonomy$Feature.ID
head(gtdb_taxonomy)

# Selecting OTUs with positive fold changes
positive_fold_change_otus <- filtered_results %>%
  filter(LogFoldChange > 0) %>%
  select(Taxon, LogFoldChange, AdjustedPValue)

head(positive_fold_change_otus)

# Selecting OTUs with negative fold changes
negative_fold_change_otus <- filtered_results %>%
  filter(LogFoldChange < 0) %>%
  select(Taxon, LogFoldChange, AdjustedPValue)

head(negative_fold_change_otus)

# Merging positive_fold_change_otus with gtdb_taxonomy to get taxonomy information
positive_fold_change_otus_with_taxonomy <- positive_fold_change_otus %>%
  left_join(gtdb_taxonomy, by = c("Taxon" = "Feature.ID"))

head(positive_fold_change_otus_with_taxonomy)
dim(positive_fold_change_otus_with_taxonomy)
positive_fold_change_otus_with_taxonomy[order(positive_fold_change_otus_with_taxonomy$LogFoldChange, decreasing = TRUE), ]

# Merging negative_fold_change_otus with gtdb_taxonomy to get taxonomy information
negative_fold_change_otus_with_taxonomy <- negative_fold_change_otus %>%
  left_join(gtdb_taxonomy, by = c("Taxon" = "Feature.ID"))
head(negative_fold_change_otus_with_taxonomy)
dim(negative_fold_change_otus_with_taxonomy)
negative_fold_change_otus_with_taxonomy[order(negative_fold_change_otus_with_taxonomy$LogFoldChange, decreasing = FALSE), ]

# Computing CPM for OTU counts
otu <- as(otu_table(phseq), "matrix")
otu_cpm <- cpm(otu, log = FALSE)

# Creating a new phyloseq object with the CPM values
otu_table(phseq) <- otu_table(otu_cpm, taxa_are_rows = TRUE)
phsig <- prune_taxa(filtered_results$Taxon, phseq)

library(ggplot2)

df_plot <- psmelt(phsig)
head(df_plot)

# Computing average OTU abundance for each group in DayPeriod
average_otu <- df_plot %>%
  group_by(DayPeriod, OTU) %>%
  summarise(AverageAbundance = mean(Abundance, na.rm = TRUE)) %>%
  ungroup()

head(average_otu)

# Compute fold change of a particular OTU between two groups in DayPeriod
otu_of_interest <- "305420"  # Replace with the OTU of interest
fold_change <- average_otu %>%
  filter(OTU == otu_of_interest) %>%
  spread(key = DayPeriod, value = AverageAbundance) %>%
  mutate(FoldChange = night / day)  # Assuming "Day" and "Night" are the group names

fold_change

# Boxplot or violin plot
ggplot(df_plot, aes(x = DayPeriod, y = Abundance, fill = DayPeriod)) +
  geom_boxplot(outlier.shape = NA, alpha = 0.6) +
  geom_jitter(width = 0.2, size = 0.8, alpha = 0.4) +
  facet_wrap(~ OTU, scales = "free_y") +
  labs(title = "Differentially Abundant OTUs between Day and Night",
       y = "Abundance (CPM)", x = "") +
  theme_minimal() +
  theme(strip.text = element_text(size = 10), axis.text.x = element_text(angle = 45, hjust = 1))

sample_info <- data.frame(sample_data(phseq))
annotation_col <- data.frame(DayPeriod = sample_info$DayPeriod)
rownames(annotation_col) <- rownames(sample_info)

library(pheatmap)

# Sorting samples by group (DayPeriod) before generating the heatmap
sorted_samples <- sample_info[order(sample_info$DayPeriod), ]
sorted_otu <- otu_table(phseq)[, rownames(sorted_samples)]

pdf("heatmap_differentially_abundant_otus.pdf", width = 10, height = 8)
pheatmap(sorted_otu,
      annotation_col = annotation_col[rownames(sorted_samples), , drop = FALSE],
      cluster_cols = FALSE,  # Disable clustering for columns to preserve group order
      cluster_rows = TRUE,
      scale = "row",  # z-score normalization across each OTU
      show_rownames = TRUE,
      show_colnames = FALSE,
      fontsize_row = 2,
      color = colorRampPalette(c("navy", "white", "firebrick3"))(100),
      main = "Heatmap of Differentially Abundant OTUs (log2 CPM)")
dev.off()

png("heatmap_differentially_abundant_otus_adjusted.png", width = 1200, height = 1000, res = 300)
pheatmap(sorted_otu,
  annotation_col = annotation_col[rownames(sorted_samples), , drop = FALSE],
  cluster_cols = FALSE,  # Disable clustering for columns to preserve group order
  cluster_rows = TRUE,
  scale = "row",  # z-score normalization across each OTU
  show_rownames = TRUE,
  show_colnames = FALSE,
  fontsize = 2,
  fontsize_row = 1,
  fontsize_col = 6,  # Decrease font size for column annotations
  legend = TRUE,
  annotation_names_col = TRUE,
  annotation_names_row = FALSE,
  color = colorRampPalette(c("navy", "white", "firebrick3"))(100),
  main = "Heatmap of Differentially Abundant OTUs (log2 CPM)")
dev.off()
```







