---
title: "Heatmaps for day and night samples in Wallace et al (2018) with paired RNA-Seq"
output: html_notebook
---

Heatmaps to visualize OTU counts for day and night samples in Wallace et al (2018) with paired RNA-Seq data.

```{r}
#install.packages("pheatmap")
library(pheatmap)
```

Importing the OTU tables for day and night samples:

```{r}
# Import the OTU table with all samples
#otu_wallace2018 <- read.csv('/home/santosrac/Projects/UGA_RACS/16S/otu_matrices/2f_otu_table.sample_filtered.no_mitochondria_chloroplast.tsv', sep='\t', header=TRUE, row.names=1)
#head(otu_wallace2018)
# Sum all values in the dataframe
#sum_otu_wallace2018 <- sum(otu_wallace2018, na.rm = TRUE)

# Total counts for OTU table with samples with a paired RNA-Seq was computed in the notebook normalize_filter_otu_abundance.ipynb
sum_otu_wallace2018 = 37828305

# Import the OTU table with day and night samples after filtering steps
day_otu_filtered_sparxcc_wallace2018 <- read.csv('/home/santosrac/Projects/UGA_RACS/IntegrationMetataxExpression/SparXCC_OTUs_paper/filtered_otu_table_day_filtered_rel_abund_cv_filtered.tsv', sep='\t', header=TRUE, row.names=1)
colnames(day_otu_filtered_sparxcc_wallace2018) <- gsub("^X", "", colnames(day_otu_filtered_sparxcc_wallace2018))
colnames(day_otu_filtered_sparxcc_wallace2018) <- gsub("^", "day_", colnames(day_otu_filtered_sparxcc_wallace2018))
night_otu_filtered_sparxcc_wallace2018 <- read.csv('/home/santosrac/Projects/UGA_RACS/IntegrationMetataxExpression/SparXCC_OTUs_paper/filtered_otu_table_night_filtered_rel_abund_cv_filtered.tsv', sep='\t', header=TRUE, row.names=1)
colnames(night_otu_filtered_sparxcc_wallace2018) <- gsub("^X", "", colnames(night_otu_filtered_sparxcc_wallace2018))
colnames(night_otu_filtered_sparxcc_wallace2018) <- gsub("^", "night_", colnames(night_otu_filtered_sparxcc_wallace2018))
head(day_otu_filtered_sparxcc_wallace2018)
head(night_otu_filtered_sparxcc_wallace2018)

# Merge the day and night dataframes by row names
merged_otu_filtered_sparxcc_wallace2018 <- merge(day_otu_filtered_sparxcc_wallace2018, night_otu_filtered_sparxcc_wallace2018, by = "row.names", all = TRUE)
rownames(merged_otu_filtered_sparxcc_wallace2018) <- merged_otu_filtered_sparxcc_wallace2018$Row.names
merged_otu_filtered_sparxcc_wallace2018 <- merged_otu_filtered_sparxcc_wallace2018[ , -1]
dim(merged_otu_filtered_sparxcc_wallace2018)
head(merged_otu_filtered_sparxcc_wallace2018)
sum_otu_wallace2018_filtered <- sum(merged_otu_filtered_sparxcc_wallace2018, na.rm = TRUE)
sum_otu_wallace2018_filtered

# Number of OTUs in respect to the OTU table of samples with a paired RNA-Seq
# (28.24M/37.8M)*100
(sum_otu_wallace2018_filtered / sum_otu_wallace2018) * 100

# Compute CPM (Counts Per Million) values for merged_otu_filtered_sparxcc_wallace2018
merged_otu_filtered_sparxcc_wallace2018_cpm <- sweep(merged_otu_filtered_sparxcc_wallace2018, 2, colSums(merged_otu_filtered_sparxcc_wallace2018, na.rm = TRUE), FUN = "/") * 1e6
head(merged_otu_filtered_sparxcc_wallace2018_cpm)
```


Heatmap with all OTUs (subset of OTUs with paired RNA-Seq data, after filtering - 276 OTUs) for day and night samples:

```{r}
# Import the GTDB taxonomy table
gtdb_taxonomy <- read.csv("/home/santosrac/Projects/UGA_RACS/16S/otu_taxonomy/gtdb_taxonomy.tsv", sep='\t', header=TRUE)
gtdb_taxonomy <- gtdb_taxonomy[, -3]
rownames(gtdb_taxonomy) <- gtdb_taxonomy$Feature.ID
head(gtdb_taxonomy)

# Filter gtdb_taxonomy based on the row names of merged_otu_filtered_sparxcc_wallace2018_cpm
filtered_gtdb_taxonomy <- gtdb_taxonomy[rownames(gtdb_taxonomy) %in% rownames(merged_otu_filtered_sparxcc_wallace2018_cpm), ]
dim(filtered_gtdb_taxonomy)
head(filtered_gtdb_taxonomy)

# Merge the OTU table with the taxonomy table
merged_otu_taxonomy <- merge(merged_otu_filtered_sparxcc_wallace2018_cpm, filtered_gtdb_taxonomy, by = "row.names", all = TRUE)
rownames(merged_otu_taxonomy) <- merged_otu_taxonomy$Row.names
merged_otu_taxonomy <- merged_otu_taxonomy[ , -1]
merged_otu_taxonomy <- merged_otu_taxonomy[, !colnames(merged_otu_taxonomy) %in% c("Feature.ID")]
dim(merged_otu_taxonomy)
head(merged_otu_taxonomy)

# Split the Taxon column into separate columns for each taxonomic level
taxon_levels <- c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species")
taxon_split <- strsplit(as.character(merged_otu_taxonomy$Taxon), ";")
taxon_df <- do.call(rbind, lapply(taxon_split, function(x) {
    length(x) <- length(taxon_levels) # Ensure all levels are present
    return(x)
}))
colnames(taxon_df) <- taxon_levels
merged_otu_taxonomy <- cbind(merged_otu_taxonomy, taxon_df)
head(merged_otu_taxonomy)

# Remove the Taxon column from merged_otu_taxonomy
merged_otu <- merged_otu_taxonomy[, !colnames(merged_otu_taxonomy) %in% c("Taxon")]
head(merged_otu)

# Create a new dataframe with only the taxonomic columns (row annotations for pheatmap)
taxonomy_df <- merged_otu[, c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species")]
head(taxonomy_df)
dim(taxonomy_df)

# Remove columns in taxonomy_df from merged_otu_taxonomy
merged_otu <- merged_otu[, !colnames(merged_otu) %in% colnames(taxonomy_df)]
head(merged_otu)
dim(merged_otu)

# Transpose the collapsed_otu_taxonomy dataframe
samples_annotation_day_period <- t(merged_otu)
samples_annotation_day_period <- as.data.frame(samples_annotation_day_period)
samples_annotation_day_period$DayPeriod <- ifelse(grepl("^night_", rownames(samples_annotation_day_period)), "night", "day")
samples_annotation_day_period <- samples_annotation_day_period[, "DayPeriod", drop = FALSE]
head(samples_annotation_day_period)
tail(samples_annotation_day_period)

# Figure with all OTUs (subset of OTUs with paired RNA-Seq data, after filtering - 276 OTUs)
pdf("all_otus.pdf", width = 12, height = 30)
pheatmap(merged_otu, scale="row", annotation_col = samples_annotation_day_period, annotation_row = taxonomy_df,
         cluster_rows = TRUE, cluster_cols = FALSE, show_colnames = FALSE)
dev.off()

png("all_otus.png", width = 3200, height = 8000, res = 300)
pheatmap(merged_otu, scale="row", annotation_col = samples_annotation_day_period, annotation_row = taxonomy_df,
         cluster_rows = TRUE, cluster_cols = FALSE, show_colnames = FALSE)
dev.off()
```

Heatmap for collapsed taxa (Order, Family, Genus) for day and night samples:

```{r}
# Remove the Taxon column from merged_otu_taxonomy
merged_otu <- merged_otu_taxonomy[, !colnames(merged_otu_taxonomy) %in% c("Taxon")]

# Replace NA values with "Unknown" in the merged_otu dataframe
merged_otu[is.na(merged_otu)] <- "Unknown"

# Collapse the OTU table by Order
merged_otu_collapsed <- aggregate(merged_otu[, 1:404], by = list(merged_otu$Order), FUN = sum)
rownames(merged_otu_collapsed) <- merged_otu_collapsed$Group.1
merged_otu_collapsed <- merged_otu_collapsed[, -1]
head(merged_otu_collapsed)
dim(merged_otu_collapsed)

png("all_orders.png", width = 3200, height = 4000, res = 300)
pheatmap(merged_otu_collapsed, scale="row", annotation_col = samples_annotation_day_period, cluster_rows = TRUE, cluster_cols = FALSE, show_colnames = FALSE)
dev.off()

pdf("all_orders.pdf", width = 12, height = 15)
pheatmap(merged_otu_collapsed, scale="row", annotation_col = samples_annotation_day_period, cluster_rows = TRUE, cluster_cols = FALSE, show_colnames = FALSE)
dev.off()

# Collapse the OTU table by Family
merged_otu_collapsed <- aggregate(merged_otu[, 1:404], by = list(merged_otu$Family), FUN = sum)
rownames(merged_otu_collapsed) <- merged_otu_collapsed$Group.1
merged_otu_collapsed <- merged_otu_collapsed[, -1]
head(merged_otu_collapsed)
dim(merged_otu_collapsed)

png("all_families.png", width = 3200, height = 5500, res = 300)
pheatmap(merged_otu_collapsed, scale="row", annotation_col = samples_annotation_day_period, cluster_rows = TRUE, cluster_cols = FALSE, show_colnames = FALSE)
dev.off()

pdf("all_families.pdf", width = 12, height = 20)
pheatmap(merged_otu_collapsed, scale="row", annotation_col = samples_annotation_day_period, cluster_rows = TRUE, cluster_cols = FALSE, show_colnames = FALSE)
dev.off()

# Collapse the OTU table by Genus
merged_otu_collapsed <- aggregate(merged_otu[, 1:404], by = list(merged_otu$Genus), FUN = sum)
rownames(merged_otu_collapsed) <- merged_otu_collapsed$Group.1
merged_otu_collapsed <- merged_otu_collapsed[, -1]
head(merged_otu_collapsed)
dim(merged_otu_collapsed)

png("all_genus.png", width = 3200, height = 5500, res = 300)
pheatmap(merged_otu_collapsed, scale="row", annotation_col = samples_annotation_day_period, cluster_rows = TRUE, cluster_cols = FALSE, show_colnames = FALSE)
dev.off()

pdf("all_genus.pdf", width = 12, height = 20)
pheatmap(merged_otu_collapsed, scale="row", annotation_col = samples_annotation_day_period, cluster_rows = TRUE, cluster_cols = FALSE, show_colnames = FALSE)
dev.off()
```



```{r}
day_network_zi_pi_metrics <- read.table("/home/santosrac/Repositories/maize_transcriptome_microbiome_networks/day_network_zi_pi_metrics.tsv",
                header = TRUE, sep = "\t")
night_network_zi_pi_metrics <- read.table("/home/santosrac/Repositories/maize_transcriptome_microbiome_networks/night_network_zi_pi_metrics.tsv",
                header = TRUE, sep = "\t")

day_connectors <- subset(day_network_zi_pi_metrics, Zi < 2.5 & Pi > 0.62)
night_connectors <- subset(night_network_zi_pi_metrics, Zi < 2.5 & Pi > 0.62)
dim(day_connectors)
dim(night_connectors)

# Filter merged_otu_filtered_sparxcc_wallace2018 based on Node columns of day_connectors
filtered_day_otu_connectors <- day_otu_filtered_sparxcc_wallace2018[rownames(day_otu_filtered_sparxcc_wallace2018) %in% day_connectors$Node, ]
filtered_night_otu_connectors <- night_otu_filtered_sparxcc_wallace2018[rownames(night_otu_filtered_sparxcc_wallace2018) %in% night_connectors$Node, ]

pdf("scaled_abundance_day_connectors.pdf")
# Scale the day OTU table
scaled_day_otu <- scale(day_otu_filtered_sparxcc_wallace2018)
# Compute mean values of scaled day OTU table
mean_scaled_day_otu <- colMeans(scaled_day_otu, na.rm = TRUE)
# Plot grey lines for all rows in scaled day OTU table, Highlight elements in Node column of day_connectors
matplot(t(scaled_day_otu[rownames(scaled_day_otu) %in% day_connectors$Node, ]), type = "l", lty = 1, col = "grey", xlab = "Samples", ylab = "Scaled Abundance", main = "Scaled Abundance of Day Connectors")
# Highlight mean values
lines(mean_scaled_day_otu, type = "l", lty = 1, col = "red", lwd = 2)
dev.off()

png("scaled_abundance_day_connectors.png")
# Scale the day OTU table
scaled_day_otu <- scale(day_otu_filtered_sparxcc_wallace2018)
# Compute mean values of scaled day OTU table
mean_scaled_day_otu <- colMeans(scaled_day_otu, na.rm = TRUE)
# Plot grey lines for all rows in scaled day OTU table, Highlight elements in Node column of day_connectors
matplot(t(scaled_day_otu[rownames(scaled_day_otu) %in% day_connectors$Node, ]), type = "l", lty = 1, col = "grey", xlab = "Samples", ylab = "Scaled Abundance", main = "Scaled Abundance of Day Connectors")
# Highlight mean values
lines(mean_scaled_day_otu, type = "l", lty = 1, col = "red", lwd = 2)
dev.off()

pdf("scaled_abundance_night_connectors.pdf")
# Scale the night OTU table
scaled_night_otu <- scale(night_otu_filtered_sparxcc_wallace2018)
# Compute mean values of scaled night OTU table
mean_scaled_night_otu <- colMeans(scaled_night_otu, na.rm = TRUE)
# Plot grey lines for all rows in scaled night OTU table, Highlight elements in Node column of night_connectors
matplot(t(scaled_night_otu[rownames(scaled_night_otu) %in% night_connectors$Node, ]), type = "l", lty = 1, col = "grey", xlab = "Samples", ylab = "Scaled Abundance", main = "Scaled Abundance of Night Connectors")
# Highlight mean values
lines(mean_scaled_night_otu, type = "l", lty = 1, col = "red", lwd = 2)
dev.off()

png("scaled_abundance_night_connectors.png")
# Scale the night OTU table
scaled_night_otu <- scale(night_otu_filtered_sparxcc_wallace2018)
# Compute mean values of scaled night OTU table
mean_scaled_night_otu <- colMeans(scaled_night_otu, na.rm = TRUE)
# Plot grey lines for all rows in scaled night OTU table, Highlight elements in Node column of night_connectors
matplot(t(scaled_night_otu[rownames(scaled_night_otu) %in% night_connectors$Node, ]), type = "l", lty = 1, col = "grey", xlab = "Samples", ylab = "Scaled Abundance", main = "Scaled Abundance of Night Connectors")
# Highlight mean values
lines(mean_scaled_night_otu, type = "l", lty = 1, col = "red", lwd = 2)
dev.off()
```
